### ISSUE TITLE
Question regarding use of bf16: I notice that the mixed precision policy is set to None when training in purebf16

### ISSUE THREAD

**Issue by PrateekMunjal** [meta-llama/llama-recipes#236]
In the following code (finetuning.py), I notice that mixed_precision is passed as None when using pure_bf16. 

` model = FSDP(
            model,
            auto_wrap_policy= my_auto_wrapping_policy if train_config.use_peft else wrapping_policy,
            cpu_offload=CPUOffload(offload_params=True) if fsdp_config.fsdp_cpu_offload else None,
            mixed_precision=mixed_precision_policy if not fsdp_config.pure_bf16 else None,
            sharding_strategy=fsdp_config.sharding_strategy,
            device_id=torch.cuda.current_device(),
            limit_all_gathers=True,
            sync_module_states=train_config.low_cpu_fsdp,
            param_init_fn=lambda module: module.to_empty(device=torch.device("cuda"), recurse=False)
            if train_config.low_cpu_fsdp and rank != 0 else None,
        )`

However, I wonder that when we are using bf16 should'nt we pass the mixed precision policy (instead of None) which is defined in policies/mixed_precision as follows:

`bfSixteen = MixedPrecision(
    param_dtype=torch.bfloat16,
    # Gradient communication precision.
    reduce_dtype=torch.bfloat16,
    # Buffer precision.
    buffer_dtype=torch.bfloat16,
    cast_forward_inputs=True,
)`

Note: I am interested to do a full parameter finetuning/ continual pretraining, so is it correct to pass mixed_precision argument as None in FSDP wrapper?

**Comment by mreso** [comment]
Hi @PrateekMunjal 
I think the difference between --pure_bf16 and --use_bf16 is that for --use_bf16 we keep a master copy of the model in fp32 while we do computations in bf16. For --pure_bf16 we convert the whole model into bfloat16 [here](https://github.com/facebookresearch/llama-recipes/blob/b7fd81c71239c67345d897c0eb6529eba076e8b8/src/llama_recipes/finetuning.py#L121) and do not keep the fp32 master copy.

Hope that answers your question.
cc @HamidShojanazeri

**Comment by PrateekMunjal** [comment]
Hi @mreso, thanks for your quick reply. Do you mind sharing where can I find this flag "--use_bf16"? I can only see the flag pure_bf16 used in the code almost everywhere.

Another question I wanted to ask is should we pass bfSixteen policy as mixed precision policy in FSDP when we want do full param finetuning in pure_bf16 setting?

```
bfSixteen = `MixedPrecision(`
    param_dtype=torch.bfloat16,
    # Gradient communication precision.
    reduce_dtype=torch.bfloat16,
    # Buffer precision.
    buffer_dtype=torch.bfloat16,
    cast_forward_inputs=True,
)

FSDP(
            model,
            ...,
            ....,
            mixed_precision=bfSixteen,
            ....
            )
```

I have this doubt because in currently in the [code](https://github.com/facebookresearch/llama-recipes/blob/main/src/llama_recipes/finetuning.py#L149) if we use pure_bf16 we do not pass any mixed precision policy.

**Comment by mreso** [comment]
Pardon me, I meant --mixed_precision. This will enable mixed precision training with fsdp. Enabling --pure_bf16 is not performing mixed precision as your model will be purely bfloat16. So for fsdp training there are three cases you can chose from:
- --mixed_precision which will perform mixed precision with bfloat16
- --mixed_precision + --fsdp_config.use_fp16 which will do mixed precision with float16
- --fsdp_config.pure_bf16 which does not perform mixed precision but converts the model into bfloat16

For single gpu training there is also --use_fp16 which uses the torch.cuda.autocast context to do mixed precision without fsdp on a single gpu.
